Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,239 @@
+        }
+
+        // lets check how many grids are between source and destination
+        // do a box intersection test between target and destination
+        // it's not very precise, but really cheap
+        var box = Box2.FromTwoPoints(sourceWorld, destWorld);
+        var grids = _mapManager.FindGridsIntersecting(mapId, box, true);
+
+        // we are only interested in grids that has some radiation blockers
+        var resGrids = grids.Where(grid => resistanceQuery.HasComponent(grid.GridEntityId)).ToArray();
+        var resGridsCount = resGrids.Length;
+
+        if (resGridsCount == 0)
+        {
+            // no grids found - so no blockers (just distance penalty)
+            return new RadiationRay(mapId, sourceUid,sourceWorld,
+                destUid,destWorld, rads);
+        }
+        else if (resGridsCount == 1)
+        {
+            // one grid found - use it for gridcast
+            return Gridcast(mapId, resGrids[0].GridEntityId, sourceUid, destUid,
+                sourceWorld, destWorld, rads, resistanceQuery);
+        }
+        else
+        {
+            // more than one grid - fallback to raycast
+            return Raycast(mapId, sourceUid, destUid, sourceWorld, destWorld,
+                dir.Normalized, dist, rads, blockerQuery);
+        }
+    }
+
+    private RadiationRay Gridcast(MapId mapId, EntityUid gridUid, EntityUid sourceUid, EntityUid destUid,
+        Vector2 sourceWorld, Vector2 destWorld, float incomingRads,
+        EntityQuery<RadiationGridResistanceComponent> resistanceQuery)
+    {
+        var visitedTiles = new List<(Vector2i, float?)>();
+        var radRay = new RadiationRay(mapId, sourceUid,sourceWorld,
+            destUid,destWorld, incomingRads)
+        {
+            Grid = gridUid,
+            VisitedTiles = visitedTiles
+        };
+
+        // if grid doesn't have resistance map just apply distance penalty
+        if (!resistanceQuery.TryGetComponent(gridUid, out var resistance))
+            return radRay;
+        var resistanceMap = resistance.ResistancePerTile;
+
+        // get coordinate of source and destination in grid coordinates
+        // todo: entity queries doesn't support interface - use it when IMapGridComponent will be removed
+        if (!TryComp(gridUid, out IMapGridComponent? grid))
+            return radRay;
+        var sourceGrid = grid.Grid.TileIndicesFor(sourceWorld);
+        var destGrid = grid.Grid.TileIndicesFor(destWorld);
+
+        // iterate tiles in grid line from source to destination
+        var line = Line(sourceGrid.X, sourceGrid.Y, destGrid.X, destGrid.Y);
+        foreach (var point in line)
+        {
+            if (resistanceMap.TryGetValue(point, out var resData))
+            {
+                radRay.Rads -= resData;
+                visitedTiles.Add((point, radRay.Rads));
+            }
+            else
+            {
+                visitedTiles.Add((point, null));
+            }
```

Comment:
AFAICT the visited tile stuff is only used when the debug overlay is active, which it generally won't be, so this is mostly just a waste of time. Maybe just have visitedTiles be nullable and only fill it in if there is an active overlay?